Add expression makers to `VirCtxt`

zgrannan opened the pull request:
This PR adds some helper functions to `VirCtxt` for creating expressions.
Later, these functions can also be used to include position data in the expression.

I'd like to merge this before doing refactoring of `ExprGenData` to include positions because doing so should reduce the size of that PR: most expressions will actually be constructed within these helper functions, so the changes should be mostly localized to `VirCtxt`, hopefully.

zgrannan committed:
Add expression makers to VirCtxt

Aurel300 commented on vir/src/context.rs:
```diff
@@ -90,23 +98,42 @@ impl<'tcx> VirCtxt<'tcx> {
 
     pub fn mk_true(&'tcx self) -> Expr<'tcx> {
-        self.alloc(ExprData::Const(self.alloc(ConstData::Bool(true))))
+        self.mk_const_expr(self.alloc(ConstData::Bool(true)))
```
Unrelated to this PR, but we should remember: these kind of nodes (`true`, `false`, `0`, `1`, …) should only need to be allocated once and re-used throughout.

JonasAlaif replied:
They shouldn't need to be allocated at all actually: they can just be `&'static`. However, with the current `Lazy` variant having `Fn(&'vir VirCtxt...` it makes `'vir` invariant and so this isn't possible. Imo we should get rid of `Lazy` (it can be replaced with a `with_let_bound_vars(|vars| ...)` where we never even need to `reify`, but let's discuss that.

Aurel300 commented:
I would expect `mk_bin_op_expr` to have the signature `(BinOpKind, Expr, Expr) -> Expr`, not `(BinOp) -> Expr`, and similarly for the other variants containing a struct.

zgrannan committed:
merge

zgrannan committed:
Refactor expression makers

zgrannan committed:
Merge remote-tracking branch 'aurel/rewrite-2023' into zgrannan/vcx-exprgen

zgrannan committed:
Continue refactor

zgrannan commented:
@Aurel300 @JonasAlaif I've refactored the `mk_*` functions accordingly, except for `PredicateApp`.

Doing this "flattening" for PredicateApp isn't as straightforward due to the arity encoding. The only ways I can think of doing it are either inelegant (different `mk_predicate_app_expr` functions for `KnownArity` and `UnknownArity`), or perhaps overly complex (an `ArgsFor` trait with an inner `Args` type, implemented for `KnownArity` and `UnknownArity`).

JonasAlaif committed:
Merge branch 'rewrite-2023' into zgrannan/vcx-exprgen

JonasAlaif committed:
Make corresponding `Struct` unconstructable outside `vir`

JonasAlaif commented:
@zgrannan ~could you check that I didn't miss any `struct *Data` which should only be constructed with a `mk_*` in making their fields `pub(crate)`~ actually there seem to still be a few `mk_*` missing to make all the structs in `vir/src/gendata.rs` unconstructable - could you add those too please (e.g. `FunctionGenData` and such, we also want to check that e.g. all the pres/posts are bool)

JonasAlaif committed:
Remove `PredicateAppData` manual construction

zgrannan committed:
More refactoring

zgrannan committed:
zap whitespace

zgrannan commented:
@JonasAlaif I've added the appropriate `mk_*` functions and made all struct fields `pub(crate)`.

> we also want to check that e.g. all the pres/posts are bool)

Agreed, but how about we do that in a different PR? This one is already a bit large.
